src/NuGet.Clients/NuGet.PackageManagement.UI/Models/NuGetProjectUpgradeWindowModel.cs

```diff
@@ -0,0 +1,231 @@
+        }
+
+        public bool HasIssues => Errors.Any() || Warnings.Any();
+
+        public IEnumerable<NuGetProjectUpgradeDependencyItem> UpgradeDependencyItems
+            => _upgradeDependencyItems ?? (_upgradeDependencyItems = GetUpgradeDependencyItems());
+
+        public IEnumerable<string> DirectDependencies => CollapseDependencies ? IncludedCollapsedPackages : AllPackages;
+
+        public IEnumerable<string> TransiviteDependencies => CollapseDependencies ? DependencyPackages : new List<string>();
+
+        private IEnumerable<string> DependencyPackages => _dependencyPackages ?? (_dependencyPackages = GetDependencyPackages());
+
+        private IEnumerable<string> AllPackages => _allPackages ?? (_allPackages = UpgradeDependencyItems.Select(d => d.Package.ToString()));
+
+        private IEnumerable<string> IncludedCollapsedPackages => _includedCollapsedPackages ?? (_includedCollapsedPackages = GetIncludedCollapsedPackages());
+
+        private IEnumerable<string> GetDependencyPackages()
+        {
+            return UpgradeDependencyItems.Where(d => d.DependingPackages.Any()).Select(d => d.ToString());
+        }
+
+        private IEnumerable<string> GetIncludedCollapsedPackages()
+        {
+            return UpgradeDependencyItems
+                .Where(upgradeDependencyItem => !upgradeDependencyItem.DependingPackages.Any())
+                .Select(upgradeDependencyItem => upgradeDependencyItem.Package.ToString());
+        }
+
+        private void InitPackageUpgradeIssues()
+        {
+            _warnings = new List<string>();
+            _errors = new List<string>();
+            _notFoundPackages = new List<PackageIdentity>();
+
+            var msBuildNuGetProject = (MSBuildNuGetProject)Project;
+            var framework = msBuildNuGetProject.ProjectSystem.TargetFramework;
+            var folderNuGetProject = msBuildNuGetProject.FolderNuGetProject;
+
+            foreach (var package in PackageDependencyInfos)
+            {
+                // We create a new PackageIdentity here, otherwise we would be passing in a PackageDependencyInfo
+                // which includes dependencies in its ToString().
+                InitPackageUpgradeIssues(folderNuGetProject, new PackageIdentity(package.Id, package.Version), framework);
+            }
+        }
+
+        private void InitPackageUpgradeIssues(FolderNuGetProject folderNuGetProject, PackageIdentity packageIdentity, NuGetFramework framework)
+        {
+            // Confirm package exists
+            var packagePath = folderNuGetProject.GetInstalledPackageFilePath(packageIdentity);
+            if (string.IsNullOrEmpty(packagePath))
+            {
+                _errors.Add(string.Format(CultureInfo.CurrentCulture, Resources.NuGetUpgradeError_CannotFindPackage, packageIdentity));
+                _notFoundPackages.Add(packageIdentity);
+            }
+            else
+            {
+                var reader = new PackageArchiveReader(packagePath);
```

Comment:
It is possible for this to not exist if the user is in nuspec only mode, that is likely rare but just make sure this fails with a helpful message about not finding the nupkg

Reply:
i digged through the code a little bit, and the migrator does a restore of missing packages before it executes, and that restore has packagesavemodes set to defaultv2. (in fact, any restore done by VS always writes the nupkg regardless of what the packagesavemode would be in the nuget.config file).